### PR TITLE
Canonicalize path when discovering repositories

### PR DESCRIPTION
**Pull request by jeffparsons**
Canonicalize the given path so that `Path::parent` _actually_ gives
us the parent directory. (`Path::parent` just strips off the last
path component, which means it will not do what you expect when
working with paths that contain '..'.)

### CODE REVIEW

**Comment by jeffparsons** [comment]
Note that previously `discover::existing` would also fail to find the repo if your process's current working directory was somewhere nested under the git repo and you pass the path to discover from as `"."`. (That's how I actually found it.)

**Review by edward-shen — COMMENTED** [review]
As general feedback on the tests, can we not just directly compare the `Result` instead of `.expect()`ing the output? 

e.g. 
```rust
assert_eq!(path.as_ref().canonicalize(), dir.canonicalize(), "desc")
```
instead of 
```rust
assert_eq!(path.as_ref().canonicalize().expect("..."), dir.canonicalize().expect("..."), "desc")
```

**Comment by jeffparsons** [comment]
@edward-shen I don't think so. Unfortunately `std::io::Error` doesn't impl `Eq`, so you can't just throw the `Result` into `assert_eq!` as-is.

I thought about adding a `canonicalize_unwrap` "extension method" just for the tests, or something like that, but I think it may actually serve to obscure what's going on overall rather than helping readability.

**Comment by jeffparsons** [comment]
The failing job looks unrelated to my changes, and I think it's failing on _main_ just the same.

```
Run cargo fmt --all -- --check
Error: `/home/runner/work/gitoxide/gitoxide/git-config` is a directory
Error: Process completed with exit code 1.
```

**Comment by edward-shen** [comment]
You're right that `std::io::Error` doesn't impl those traits. Looks like we return a `Result` in those tests, can we just use `path.as_ref()?` then?

As for CI errors, yeah that's on me, need to investigate.

**Comment by jeffparsons** [comment]
> Looks like we return a `Result` in those tests, can we just use `path.as_ref()?` then?

I'm not sure I follow. We're comparing two paths to each other in each test, so I don't see what `Result` I could return from the test.

File: git-repository/tests/discover/mod.rs [hunk 1]
```diff
@@ -31,8 +35,10 @@ mod existing {
         let path = git_repository::path::discover::existing(&dir)?;
         assert_eq!(path.kind(), Kind::WorkingTree);
         assert_eq!(
-            path.into_repository_directory(),
-            dir,
+            path.into_repository_directory()
+                .canonicalize()
+                .expect("should be a directory"),
+            dir.canonicalize().expect("should be a directory"),
```
Comment (edward-shen) [inline git-repository/tests/discover/mod.rs:41]: Something like this:

```suggestion
            path.into_repository_directory().canonicalize()?,
            dir.canonicalize()?,
```

Comment (jeffparsons) [reply git-repository/tests/discover/mod.rs:41]: Ah, yep, sure. Happy to make that change.
I'm on mobile now; will do it when I get a chance a bit later today.

And thanks for looking into the CI issue. 🙂

**Comment by jeffparsons** [comment]
@edward-shen I've made that change and rebased on main. Should be good to go I think, but blocked on the workflow approval thing for now.

**Comment by edward-shen** [comment]
Thanks for the contribution!

**Comment by Byron** [comment]
Many thanks also from my side. Having just returned from vacation I am a bit late to the party and took the liberty to make a few adjustments. They should show up in this PR in a moment, but in short paths will now only be canonicalized when needed.

**Comment by jeffparsons** [comment]
@Byron Awesome! I thought about doing something similar but worried it might be considered overkill. 😅

This definitely feels like having our cake and eating it, too.

**Comment by Byron** [comment]
> @Byron Awesome! I thought about doing something similar but worried it might be considered overkill. 😅

Hehe, I think quality will always be appreciated, and I agree that often enough it's not exactly clear what tradeoffs to make.

If you are using `gitoxide` for a project of yours you are welcome to start a discussion to see how feasible this is right now or which contributions are needed to make it feasible.

**Pull request merged**